src/store/token/actions.js

```diff
@@ -203,29 +204,86 @@ export function requestUBIPayout(payout) {
   };
 }
 
-async function loopTransfer(from, to, value, paymentNote) {
-  return await waitAndRetryOnFail(
-    () => {
-      return core.token.transfer(from, to, value, paymentNote);
-    },
-    () => {
-      return true;
-    },
-    {},
-    () => {
-      return core.token.updateTransferSteps(from, to, value);
-    },
-  );
+// Recursive helper function for transfer
+async function loopTransfer(
+  from,
+  to,
+  value,
+  paymentNote,
+  hops,
+  attemptsLeft,
+  errorsMessages = '',
+) {
+  if (attemptsLeft === 0 || hops === 0) {
+    // ran out of attempts or hops, cannot attempt further transfers
+    throw new TransferError(errorsMessages);
+  }
+  try {
+    return await core.token.transfer(from, to, value, paymentNote, hops);
+  } catch (error) {
+    // RETRY when path is not found or is too long
+    // UPDATE edges database when the path fails and then retry
+    // GIVE UP for other errors
+    // ---
+    // no path or path too long
+    if (
+      error.name === 'TransferError' &&
+      (error.code === ErrorCodes.TOO_COMPLEX_TRANSFER || // too many steps in found path
+        error.code === ErrorCodes.UNKNOWN_ERROR || // includes timeout error from api
+        error.code === ErrorCodes.INVALID_TRANSFER) // other errors from find transitive transfer
+    ) {
+      // retry with fewer hops
+      return await loopTransfer(
+        from,
+        to,
+        value,
+        paymentNote,
+        hops - 1,
+        attemptsLeft - 1,
+        errorsMessages.concat(' ', error.message),
+      );
+    }
+    // if the path is found but it is invalid
+    else if (
+      (error.name === 'TransferError' && ErrorCodes.TRANSFER_NOT_FOUND) || // search complete and no path found
+      error.name !== 'TransferError' // includes errors from attempting transfer with an invalid path
+    ) {
+      // update the edges db for trust-adjacent safes
+      await core.token.updateTransferSteps(from, to, value, hops);
+      // try again after update with the same parameters
+      return await loopTransfer(
+        from,
+        to,
+        value,
+        paymentNote,
+        hops,
+        attemptsLeft - 1,
+        errorsMessages.concat(' ', error.message),
+      );
+    }
+    // any other errors will result in propagating the error
+    else {
+      throw error;
```

Comment:
what about information for users? shouldn't we display a message for them?

Reply:
If you look where this error is caught (i think in SendConfirm) it will always display something like failed transaction. I kept these as a debug log as the error will go in the logs

Reply:
Is interpretation for the user there?

Reply:
Yes and it will be improved when working on the task to improve that view